### ISSUE TITLE
'property' in Object not detected in strictNullChecks

### ISSUE THREAD

**Issue by kitsonk** [Microsoft/TypeScript#8396]
**TypeScript Version:** 

nightly (1.9.0-dev.20160430)

**Code**

```ts
function foo(options?: { foo?: { bar: string; }; }) {
	if (options && 'foo' in options) {
		console.log(options.foo.bar); // <-- Object is possibly 'undefined'.
	}
};
```

**Expected behavior:**

No errors.

**Actual behavior:** 

With `strictNullChecks = true` it is reported that the object is possibly `undefined`.  While `options && options.foo` will guard against undefined, it is a common pattern to use `'property' in object` as a similar guard.

**Comment by weswigham** [comment]
This would only guard against `undefined`, not `null`, right?

**Comment by kitsonk** [comment]
Correct, but this also could potentially suffer from not detecting `null` as well:

```typescript
function foo(options?: { foo?: { bar: string; }; }) {
	if (options && typeof options.foo === 'object') {
		console.log(options.foo.bar); // <-- no error with strict null checks
	}
};
```

And both are common patterns.  While it would be possible to easily "break" this at runtime in just JavaScript, I think in TypeScript, unless you have people performing type abuse, then it is unlikely to be consumed in a way where both patterns are sufficient.

**Comment by ivogabe** [comment]
I don't think it guards against `undefined` or `null`, but it guards against the existence. So `{ foo?: { bar: string; }; }` becomes `{ foo: { bar: string; }; }` (question mark removed).

**Comment by kitsonk** [comment]
Again, correct:

```typescript
foo({}); // <-- guarded and TypeScript valid

foo({ foo: undefined }); // <-- not guarded, but TypeScript error

foo({ foo: null }); // <-- not guarded, but TypeScript error
```

But the last one also isn't guarded against with `typeof options.foo === 'object'` but it passes the `strictNullChecks`.

**Comment by weswigham** [comment]
```ts
var x = {}
x.foo = undefined;
'foo' in x // true
```

It doesn't even guard against undefined?

**Comment by kitsonk** [comment]
And:

```js
var x = {};
x.foo = null;
typeof x.foo === 'object'; // true
```

It is either too permissive or too conservative.  I vote for too conservative.  The intent is clear.

**Comment by weswigham** [comment]
@kitsonk  That's because `null` is, explicitly, an object type.

**Comment by ivogabe** [comment]
But `foo` can never be `undefined` or `null`, according to the type `{ foo?: { bar: string; }; }`. The property does either not exist, or has a value that is not undefined or null (under strictNullChecks). `in` checks for the existence of the property, so guarantees that the value will not be `undefined` when you access it.

**Comment by kitsonk** [comment]
> That's because null is, explicitly, an object type.

Yes, my point is `typeof x.foo === 'object'` passes `strictNullChecks` when it can actually be `null`, so why doesn't `'foo' in x` also pass `strictNullChecks` when it signifies that there is an actual property which is defined on `x`.  Either both should fail or both should pass, let alone the pattern `'foo' in x' being a common in the wild guard for an existence of a property (which is a logical guard if you have an optional property in a type on an object).

**Comment by ivogabe** [comment]
What I just said isn't true, apparently. The following compiles, though I would expect that this would give an error:
```typescript
interface A { x?: string }
const b: A = { x: undefined };
```
Could this behavior be changed? Currently `{ x?: string }` and `{ x: string | undefined }` are the same types.

**Comment by weswigham** [comment]
@ivogabe I'm not sure the distinction is so clear - many developers conflate an `undefined` field with a 'never set' field. (I mean, the value _is_ called `undefined`) TS Compiler include.

**Comment by weswigham** [comment]
Anyways, the `in` operator (in the true branch) doesn't allow for any type narrowing - the false branch, however, could be used to guarantee that a field doesn't exist (and is therefore exactly `undefined`).

So, back in your original example:
```ts
function foo(options?: { foo?: { bar: string; }; }) {
    if (options && 'foo' in options) {
        console.log(options.foo.bar); // <-- Object is possibly 'undefined'.
    }
};
```

The compiler is right. The `in` operator doesn't tell us that the value of `bar` is not `undefined`.

**Comment by kitsonk** [comment]
Ok, fair enough...

Given that:

```typescript
foo({ foo: undefined }); // <- no TypeScript error, but silly and surprising
foo({ foo: null }); // <- TypeScript error
```

**Comment by malibuzios** [comment]
It seems like the `in` operator matches the behavior of `for .. in`, where it would also include properties explicitly set to `undefined`.

From [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in):
>If you set a property to undefined but do not delete it, the in operator returns true for that property.

I must say that makes it quite unreliable, almost to the point of being completely useless in practice. Having the property name as string is also not very desirable for TypeScript, since it doesn't (currently) respond to refactoring operations like rename, find all references etc.
